src/server/req_delete.c

```diff
@@ -343,88 +381,54 @@ req_deletejob(struct batch_request *preq)
 	else
 		qdel_mail = 1;
-	parent = chk_job_request(jid, preq, &jt);
-	if (parent == NULL)
+
+	if (FALSE == chk_job_request(jid, preq, &jt, &js, &pjob, &parent, &offset))
 		return; /* note, req_reject already called */
 
 	if (delhist) {
-		rc = check_deletehistoryjob(preq);
+		rc = check_deletehistoryjob(preq, pjob, parent, jt);
 		if (rc == TRUE)
 			return;
 	}
 
 
-	if (jt == IS_ARRAY_NO) {
+	if ((jt == IS_ARRAY_NO) || (jt == IS_ARRAY_Single)) {
 
-		/* just a regular job, pass it on down the line and be done
+		/* just a regular job or a subjob, pass it on down the line and be done
 		 * If the request if to purge the history of the job then set ji_deletehistory to 1
 		 */
-		if (delhist)
-			parent->ji_deletehistory = 1;
-		req_deletejob2(preq, parent);
-		return;
+		if (pjob != NULL) {
+			if (delhist)
+				pjob->ji_deletehistory = 1;
+			req_deletejob2(preq, pjob);
 
-	} else if (jt == IS_ARRAY_Single) {
+			if (jt == IS_ARRAY_Single) {
 
-		/* single subjob, if running do full delete, */
-		/* if not then just set it expired		 */
+				/* single subjob, update tracking table header count */
 
-		offset = subjob_index_to_offset(parent, get_index_from_jid(jid));
-		if (offset == -1) {
-			req_reject(PBSE_UNKJOBID, 0, preq);
-			return;
-		}
-		i = get_subjob_state(parent, offset);
-		if (i == -1) {
-			req_reject(PBSE_IVALREQ, 0, preq);
-			return;
-		}
-		/*
-		 * If the request if to also purge the history of the sub job then log a message since
-		 * server does not store the history of subjobs
-		 */
-		if (delhist)
-			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB,
-				LOG_INFO, jid,
-				msg_nosave_histarrayjob);
-
-		if (i == JOB_STATE_RUNNING) {
-			pjob = find_job(jid); /* get ptr to the subjob */
-			if (pjob) {
-				req_deletejob2(preq, pjob);
 				if (parent->ji_ajtrk)
 					if (pjob->ji_terminated)
 						parent->ji_ajtrk->tkm_dsubjsct++;
+
+				chk_array_doneness(parent);
+			}
+			return;
+		} else {
+			if (js == JOB_STATE_EXPIRED) {
+				req_reject(PBSE_BADSTATE, 0, preq);
+				return;
 			} else {
+				acct_del_write(jid, parent, preq, 0);
 				parent->ji_ajtrk->tkm_tbl[offset].trk_substate =
 					JOB_SUBSTATE_TERMINATED;
-				set_subjob_tblstate(pjob, offset, JOB_STATE_EXPIRED);
+				set_subjob_tblstate(parent, offset, JOB_STATE_EXPIRED);
 				parent->ji_ajtrk->tkm_dsubjsct++;
-				acct_del_write(jid, pjob, preq, 0);
-			}
-		} else if (((i == JOB_STATE_EXITING) && (forcedel == 0)) ||
-			(i == JOB_STATE_EXPIRED)) {
-			req_reject(PBSE_BADSTATE, 0, preq);
-			return;
-		} else {
-			acct_del_write(jid, parent, preq, 0);
-			parent->ji_ajtrk->tkm_tbl[offset].trk_substate =
-				JOB_SUBSTATE_TERMINATED;
-			set_subjob_tblstate(parent, offset, JOB_STATE_EXPIRED);
-			parent->ji_ajtrk->tkm_dsubjsct++;
 
-			if (i == JOB_STATE_EXITING) {
-				pjob = find_job(jid); /* subjob */
-				if (pjob) {
-					discard_job(pjob, "Forced Delete", 1);
-					rel_resc(pjob);
-					job_purge(pjob);
-				}
+				reply_ack(preq);
```

Comment:
Earlier if a qdel without force is issued we used to reject the request if the state of job is in JOB_STATE_EXITING or JOB_STATE_EXPIRED. Now since you removed this block of code qdel request is accepted during these cases. Please modify the code and handle it accordingly.

Reply:
It will be handled in req_deletejob2()